linkerd/meshtls/src/client.rs

```diff
@@ -0,0 +1,200 @@
+use linkerd_io as io;
+use linkerd_stack::{NewService, Service};
+use linkerd_tls::{ClientTls, HasNegotiatedProtocol, NegotiatedProtocolRef};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[cfg(feature = "rustls")]
+use crate::rustls;
+
+#[derive(Clone, Debug)]
+pub enum NewClient {
+    #[cfg(feature = "rustls")]
+    Rustls(rustls::NewClient),
+}
+
+#[derive(Clone)]
+pub enum Connect {
+    #[cfg(feature = "rustls")]
+    Rustls(rustls::Connect),
+}
+
+#[pin_project::pin_project(project = ConnectFutureProj)]
+pub enum ConnectFuture<I> {
+    #[cfg(feature = "rustls")]
+    Rustls(#[pin] rustls::ConnectFuture<I>),
+}
+
+#[pin_project::pin_project(project = ClientIoProj)]
+#[derive(Debug)]
+pub enum ClientIo<I> {
+    #[cfg(feature = "rustls")]
+    Rustls(#[pin] rustls::ClientIo<I>),
+}
+
+// === impl NewClient ===
+
+impl NewService<ClientTls> for NewClient {
+    type Service = Connect;
+
+    fn new_service(&self, target: ClientTls) -> Self::Service {
+        #[cfg(feature = "rustls")]
+        if let Self::Rustls(new_client) = self {
+            return Connect::Rustls(new_client.new_service(target));
+        }
+
+        unreachable!()
```

Comment:
nit/TIOLI: not a big deal, but maybe this should say something like
```suggestion
        unreachable!("attempted o initiate a mTLS connection, but no TLS backends were enabled at compile time!")
```
or something, _just in case_ someone makes a build of the proxy where no backends were enabled?

Reply:
alternatively, we could probably make this unreachable at compile-time by putting `#[cfg(any(feature = "rustls", feature = "boring", ...etc))]` on the `Service` impl (and other stuff in this crate), but i'm not sure if that would make other stuff more complicated...

Reply:
I changed this to use a match for all of these so there's no longer an unreachable case. I've added a build.rs check that ensures at least one feature is enabled.